Protobuf Extension task failed

### Affected Version

v0.14.2

### Description

Just follow steps described here: https://druid.apache.org/docs/latest/development/extensions-core/protobuf.html

Got exception in Ingestion Task: 
```
java.lang.NoSuchMethodError: com.github.os72.protobuf.dynamic.DynamicSchema.getMessageDescriptor(Ljava/lang/String;)Lshaded/com/google/protobuf/Descriptors$Descriptor;
	at org.apache.druid.data.input.protobuf.ProtobufInputRowParser.getDescriptor(ProtobufInputRowParser.java:166) ~[?:?]
	at org.apache.druid.data.input.protobuf.ProtobufInputRowParser.initDescriptor(ProtobufInputRowParser.java:90) ~[?:?]
	at org.apache.druid.data.input.protobuf.ProtobufInputRowParser.parseBatch(ProtobufInputRowParser.java:101) ~[?:?]
	at org.apache.druid.data.input.protobuf.ProtobufInputRowParser.parseBatch(ProtobufInputRowParser.java:52) ~[?:?]
	at org.apache.druid.segment.transform.TransformingInputRowParser.parseBatch(TransformingInputRowParser.java:50) ~[druid-processing-0.14.2-incubating.jar:0.14.2-incubating]
	at org.apache.druid.indexing.seekablestream.SeekableStreamIndexTaskRunner.runInternal(SeekableStreamIndexTaskRunner.java:562) ~[druid-indexing-service-0.14.2-incubating.jar:0.14.2-incubating]
	at org.apache.druid.indexing.seekablestream.SeekableStreamIndexTaskRunner.run(SeekableStreamIndexTaskRunner.java:246) ~[druid-indexing-service-0.14.2-incubating.jar:0.14.2-incubating]
	at org.apache.druid.indexing.seekablestream.SeekableStreamIndexTask.run(SeekableStreamIndexTask.java:166) ~[druid-indexing-service-0.14.2-incubating.jar:0.14.2-incubating]
	at org.apache.druid.indexing.overlord.SingleTaskBackgroundRunner$SingleTaskBackgroundRunnerCallable.call(SingleTaskBackgroundRunner.java:419) [druid-indexing-service-0.14.2-incubating.jar:0.14.2-incubating]
	at org.apache.druid.indexing.overlord.SingleTaskBackgroundRunner$SingleTaskBackgroundRunnerCallable.call(SingleTaskBackgroundRunner.java:391) [druid-indexing-service-0.14.2-incubating.jar:0.14.2-incubating]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_212]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_212]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_212]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_212]
Error!
java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.lang.NoSuchMethodError: com.github.os72.protobuf.dynamic.DynamicSchema.getMessageDescriptor(Ljava/lang/String;)Lshaded/com/google/protobuf/Descriptors$Descriptor;
	at com.google.common.base.Throwables.propagate(Throwables.java:160)
	at org.apache.druid.indexing.worker.executor.ExecutorLifecycle.join(ExecutorLifecycle.java:214)
	at org.apache.druid.cli.CliPeon.run(CliPeon.java:364)
	at org.apache.druid.cli.Main.main(Main.java:118)
Caused by: java.util.concurrent.ExecutionException: java.lang.NoSuchMethodError: com.github.os72.protobuf.dynamic.DynamicSchema.getMessageDescriptor(Ljava/lang/String;)Lshaded/com/google/protobuf/Descriptors$Descriptor;
	at com.google.common.util.concurrent.AbstractFuture$Sync.getValue(AbstractFuture.java:299)
	at com.google.common.util.concurrent.AbstractFuture$Sync.get(AbstractFuture.java:286)
	at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:116)
	at org.apache.druid.indexing.worker.executor.ExecutorLifecycle.join(ExecutorLifecycle.java:211)
	... 2 more
```

The reason may be that protobuf extension shaded and relocated the `com.google.protobuf` package, while the binary distribution provides the original `protobuf-dynamic-0.9.3.jar` which depends on `com.google.protobuf`.

When we remove `protobuf-dynamic-0.9.3.jar` from the druid-protobuf-extensions folder, it works without exceptions

--- a/docs/development/extensions-core/protobuf.md
+++ b/docs/development/extensions-core/protobuf.md
@@ -30,16 +30,16 @@ for [stream ingestion](../../ingestion/index.md#streaming). See corresponding do
 
 ## Example: Load Protobuf messages from Kafka
 
-This example demonstrates how to load Protobuf messages from Kafka.  Please read the [Load from Kafka tutorial](../../tutorials/tutorial-kafka.md) first.  This example will use the same "metrics" dataset.
+This example demonstrates how to load Protobuf messages from Kafka.  Please read the [Load from Kafka tutorial](../../tutorials/tutorial-kafka.md) first, and see [Kafka Indexing Service](./kafka-ingestion.md) documentation for more details.
 
-Files used in this example are found at `./examples/quickstart/protobuf` in your Druid directory.
+The files used in this example are found at [`./examples/quickstart/protobuf` in your Druid directory](https://github.com/apache/druid/tree/master/examples/quickstart/protobuf).
 
-- We will use [Kafka Indexing Service](./kafka-ingestion.md).
-- Kafka broker host is `localhost:9092`.
-- Kafka topic is `metrics_pb` instead of `metrics`.
-- datasource name is `metrics-kafka-pb` instead of `metrics-kafka` to avoid the confusion.
+For this example:
+- Kafka broker host is `localhost:9092`
+- Kafka topic is `metrics_pb`
+- Datasource name is `metrics-protobuf`
 
-Here is the metrics JSON example.
+Here is a JSON example of the 'metrics' data schema used in the example.
 
 ```json
 {
@@ -56,7 +56,7 @@ Here is the metrics JSON example.
 
 ### Proto file
 
-The proto file should look like this.  Save it as metrics.proto.
+The corresponding proto file for our 'metrics' dataset looks like this.
 
 ```
 syntax = "proto3";
@@ -74,28 +74,27 @@ message Metrics {
 
 ### Descriptor file
 
-Using the `protoc` Protobuf compiler to generate the descriptor file.  Save the metrics.desc file either in the classpath or reachable by URL.  In this example the descriptor file was saved at /tmp/metrics.desc.
+Next, we use the `protoc` Protobuf compiler to generate the descriptor file and save it as `metrics.desc`. The descriptor file must be either in the classpath or reachable by URL.  In this example the descriptor file was saved at `/tmp/metrics.desc`, however this file is also available in the example files.
 
 ```
 protoc -o /tmp/metrics.desc metrics.proto
 ```
 
-### Supervisor spec JSON
+## Create Kafka Supervisor
 
 Below is the complete Supervisor spec JSON to be submitted to the Overlord.
-Please make sure these keys are properly configured for successful ingestion.
+Make sure these keys are properly configured for successful ingestion.
 
-- `descriptor` for the descriptor file URL.
-- `protoMessageType` from the proto definition.
-- parseSpec `format` must be `json`.
-- `topic` to subscribe.  The topic is "metrics_pb" instead of "metrics".
-- `bootstrap.server` is the Kafka broker host.
+Important superivisor properties
+- `descriptor` for the descriptor file URL
+- `protoMessageType` from the proto definition
+- `parser` should have `type` set to `protobuf`, but note that the `format` of the `parseSpec` must be `json`
 
 ```json
 {
   "type": "kafka",
   "dataSchema": {
-    "dataSource": "metrics-kafka2",
+    "dataSource": "metrics-protobuf",
     "parser": {
       "type": "protobuf",
       "descriptor": "file:///tmp/metrics.desc",
@@ -165,54 +164,74 @@ Please make sure these keys are properly configured for successful ingestion.
 }
 ```
 
-## Kafka Producer
+## Adding protobuf messages to Kafka
 
-Here is the sample script that publishes the metrics to Kafka in Protobuf format.
+If necessary, from your Kafka installation directory run the following command to create the kafka topic
 
-1. Run `protoc` again with the Python binding option.  This command generates `metrics_pb2.py` file.
- ```
-  protoc -o metrics.desc metrics.proto --python_out=.
- ```
+```
+./bin/kafka-topics.sh --create --zookeeper localhost:2181 --replication-factor 1 --partitions 1 --topic metrics_pb
+```
 
-2. Create Kafka producer script.
+This example script requires `protobuf` and `kafka-python` modules. With the topic in place, messages can be inserted
 
-This script requires `protobuf` and `kafka-python` modules.
+```
+../bin/generate-example-metrics | ./pb_publisher.py
+```
+
+You can confirm that data has been inserted to your Kafka topic using the following command from your Kafka installation directory
+
+```
+./bin/kafka-console-consumer --zookeeper localhost --topic metrics_pb
+```
+
+which should print messages like this
+
+```
+millisecondsGETR"2017-04-06T03:23:56Z*2002/list:request/latencyBwww1.example.com
+```
+
+If your supervisor created in the previous step is running, the indexing tasks should begin producing the messages and the data will soon be available for querying in Druid.
+
+## Generating the example files
+
+The files provided in the example quickstart can be generated in the following manner starting with only `metrics.proto`.
+
+### `metrics.desc`
+
+The descriptor file is generated using `protoc` Protobuf compiler. Given a `.proto` file, a `.desc` file can be generated like so.
+
+```
+protoc -o metrics.desc metrics.proto
+```
+
+### `metrics_pb2.py`
+`metrics_pb2.py` is also generated with `protoc`
+
+```
+ protoc -o metrics.desc metrics.proto --python_out=.
+```
+
+### `pb_publisher.py`
+After `metrics_pb2.py` is generated, another script can be constructed to parse JSON data, convert it to protobuf, and produce to a Kafka topic
 
 ```python
-#!/usr/bin/env python
-
 import sys
 import json
 
 from kafka import KafkaProducer
 from metrics_pb2 import Metrics
 
+
 producer = KafkaProducer(bootstrap_servers='localhost:9092')
 topic = 'metrics_pb'
-metrics = Metrics()
 
 for row in iter(sys.stdin):
     d = json.loads(row)
+    metrics = Metrics()
     for k, v in d.items():
         setattr(metrics, k, v)
     pb = metrics.SerializeToString()
     producer.send(topic, pb)
-```
 
-3. run producer
-
-```
-./bin/generate-example-metrics | ./pb_publisher.py
-```
-
-4. test
-
-```
-kafka-console-consumer --zookeeper localhost --topic metrics_pb
-```
-
-It should print messages like this
-
-```
-millisecondsGETR"2017-04-06T03:23:56Z*2002/list:request/latencyBwww1.example.com
+producer.flush()
 ```

--- a/docs/development/extensions-core/protobuf.md
+++ b/docs/development/extensions-core/protobuf.md
@@ -85,7 +85,7 @@ protoc -o /tmp/metrics.desc ./quickstart/protobuf/metrics.proto
 Below is the complete Supervisor spec JSON to be submitted to the Overlord.
 Make sure these keys are properly configured for successful ingestion.
 
-Important superivisor properties
+Important supervisor properties
 - `descriptor` for the descriptor file URL
 - `protoMessageType` from the proto definition
 - `parser` should have `type` set to `protobuf`, but note that the `format` of the `parseSpec` must be `json`
@@ -164,9 +164,9 @@ Important superivisor properties
 }
 ```
 
-## Adding protobuf messages to Kafka
+## Adding Protobuf messages to Kafka
 
-If necessary, from your Kafka installation directory run the following command to create the kafka topic
+If necessary, from your Kafka installation directory run the following command to create the Kafka topic
 
 ```
 ./bin/kafka-topics.sh --create --zookeeper localhost:2181 --replication-factor 1 --partitions 1 --topic metrics_pb
@@ -212,7 +212,7 @@ protoc -o metrics.desc metrics.proto
 ```
 
 ### `pb_publisher.py`
-After `metrics_pb2.py` is generated, another script can be constructed to parse JSON data, convert it to protobuf, and produce to a Kafka topic
+After `metrics_pb2.py` is generated, another script can be constructed to parse JSON data, convert it to Protobuf, and produce to a Kafka topic
 
 ```python
 import sys

--- a/docs/development/extensions-core/protobuf.md
+++ b/docs/development/extensions-core/protobuf.md
@@ -74,10 +74,10 @@ message Metrics {
 
 ### Descriptor file
 
-Next, we use the `protoc` Protobuf compiler to generate the descriptor file and save it as `metrics.desc`. The descriptor file must be either in the classpath or reachable by URL.  In this example the descriptor file was saved at `/tmp/metrics.desc`, however this file is also available in the example files.
+Next, we use the `protoc` Protobuf compiler to generate the descriptor file and save it as `metrics.desc`. The descriptor file must be either in the classpath or reachable by URL.  In this example the descriptor file was saved at `/tmp/metrics.desc`, however this file is also available in the example files. From your Druid install directory:
 
 ```
-protoc -o /tmp/metrics.desc metrics.proto
+protoc -o /tmp/metrics.desc ./quickstart/protobuf/metrics.proto
 ```
 
 ## Create Kafka Supervisor
@@ -172,10 +172,10 @@ If necessary, from your Kafka installation directory run the following command t
 ./bin/kafka-topics.sh --create --zookeeper localhost:2181 --replication-factor 1 --partitions 1 --topic metrics_pb
 ```
 
-This example script requires `protobuf` and `kafka-python` modules. With the topic in place, messages can be inserted
+This example script requires `protobuf` and `kafka-python` modules. With the topic in place, messages can be inserted running the following command from your Druid installation directory
 
 ```
-../bin/generate-example-metrics | ./pb_publisher.py
+./bin/generate-example-metrics | ./quickstart/protobuf/pb_publisher.py
 ```
 
 You can confirm that data has been inserted to your Kafka topic using the following command from your Kafka installation directory

--- a/docs/development/extensions-core/protobuf.md
+++ b/docs/development/extensions-core/protobuf.md
@@ -215,6 +215,8 @@ protoc -o metrics.desc metrics.proto
 After `metrics_pb2.py` is generated, another script can be constructed to parse JSON data, convert it to Protobuf, and produce to a Kafka topic
 
 ```python
+#!/usr/bin/env python
+
 import sys
 import json
 

--- a/examples/quickstart/protobuf/pb_publisher.py
+++ b/examples/quickstart/protobuf/pb_publisher.py
@@ -32,3 +32,5 @@ for row in iter(sys.stdin):
         setattr(metrics, k, v)
     pb = metrics.SerializeToString()
     producer.send(topic, pb)
+
+producer.flush()

--- a/extensions-core/protobuf-extensions/pom.xml
+++ b/extensions-core/protobuf-extensions/pom.xml
@@ -109,7 +109,7 @@
         <artifactId>maven-shade-plugin</artifactId>
         <version>3.2.1</version>
         <configuration>
-          <createDependencyReducedPom>false</createDependencyReducedPom>
+          <createDependencyReducedPom>true</createDependencyReducedPom>
           <relocations>
             <relocation>
               <pattern>com.google.protobuf</pattern>

